src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -225,48 +177,183 @@ void BaseCouplingScheme::finalize()
 {
   PRECICE_TRACE();
   checkCompletenessRequiredActions();
-  PRECICE_CHECK(isInitialized(), "Called finalize() before initialize()!");
+  PRECICE_ASSERT(_isInitialized, "Called finalize() before initialize().");
+}
+
+void BaseCouplingScheme::initialize(double startTime, int startTimeWindow)
+{
+  // Initialize uses the template method pattern (https://en.wikipedia.org/wiki/Template_method_pattern).
+  PRECICE_TRACE(startTime, startTimeWindow);
+  PRECICE_ASSERT(not isInitialized());
+  PRECICE_ASSERT(math::greaterEquals(startTime, 0.0), startTime);
+  PRECICE_ASSERT(startTimeWindow >= 0, startTimeWindow);
+  _time          = startTime;
+  _timeWindows   = startTimeWindow;
+
+  if (isImplicitCouplingScheme()) {
+    if (not doesFirstStep()) {
+      PRECICE_ASSERT(not _convergenceMeasures.empty(), "Implicit scheme must have at least one convergence measure.");
+      mergeData();                             // merge send and receive data for all pp calls
+      setupConvergenceMeasures();              // needs _couplingData configured
+      setupDataMatrices(getAcceleratedData()); // Reserve memory and initialize data with zero
+      if (getAcceleration()) {
+        getAcceleration()->initialize(getAcceleratedData()); // Reserve memory, initialize
+      }
+    }
+    requireAction(constants::actionWriteIterationCheckpoint());
+    initializeTXTWriters();
+  }
+
+  initializeImplementation();
+
+  if (sendsInitializedData()) {
+    requireAction(constants::actionWriteInitialData());
+  }
+
+  _isInitialized = true;
+}
+
+void BaseCouplingScheme::initializeData()
+{
+  // InitializeData uses the template method pattern (https://en.wikipedia.org/wiki/Template_method_pattern).
+  PRECICE_CHECK(_isInitialized, "initializeData() can be called after initialize() only.");
+  PRECICE_CHECK(not _initializeDataHasBeenCalled, "initializeData() can only be called once.");
+  _initializeDataHasBeenCalled = true;
+  PRECICE_TRACE("initializeData()");
+
+  if (not _sendsInitializedData && not _receivesInitializedData) {
+    PRECICE_INFO("initializeData is skipped since no data has to be initialized.");
+    return;
+  }
+
+  PRECICE_DEBUG("Initializing Data ...");
+
+  PRECICE_CHECK(not(_sendsInitializedData && isActionRequired(constants::actionWriteInitialData())),
+                "InitialData has to be written to preCICE before calling initializeData().");
+
+  _hasDataBeenExchanged = false;
+
+  exchangeInitialData();
+}
+
+void BaseCouplingScheme::advance()
+{
+  PRECICE_TRACE(_timeWindows, _time);
+  checkCompletenessRequiredActions();
+  PRECICE_ASSERT(_isInitialized, "Before calling advance() coupling scheme has to be initialized via initialize().");
```

Comment:
Why is this an assertion and similar checks in `initializeData` real checks? We need one place where we check for false API usage. Probably easier to understand in `SolverInterfaceImpl`.

Reply:
This `PRECICE_ASSERT` was already part of `BaseCouplingScheme::initialize` before (see [here](https://github.com/precice/precice/blob/04bef9d14afdc94229f6788b2e29be5d465f9abd/src/cplscheme/SerialCouplingScheme.cpp#L39)).

I totally agree that this check should be moved into `SolverInterfaceImpl`, but I am not sure whether we should also add this additional change into this PR. It will make this PR even more complicated than it already is. 

Note that currently the changes in ` src/precice/impl/SolverInterfaceImpl.cpp` are quite minimal, which is a good thing.

Reply:
OK, could you please open an issue to clean up these checks?

Reply:
Done: https://github.com/precice/precice/issues/714